File: 2023/day17.scala
package avborup.aoc2023.day17

import collection.mutable.PriorityQueue
import collection.mutable.Map as MutableMap
import scala.collection.mutable.ArrayBuffer

object Day17 {
  def main(args: Array[String]) =
    val input = Input.read()
    println(s"Part 1: ${part1(input)}")
    println(s"Part 2: ${part2(input)}")

  def part1(input: Input) =
    dijkstra(input.start, input.dest, input.adj(maxSteps = 3))

  def part2(input: Input) =
    dijkstra(input.start, input.dest, input.adj(minSteps = 4, maxSteps = 10))
}

type Coord = (Int, Int)
type Direction = (Int, Int)
type Vertex = (Coord, Direction, Int)

case class Input(grid: Array[Array[Int]], w: Int, h: Int) {
  def start = (0, 0)
  def dest = (h - 1, w - 1)

  def adj(minSteps: Int = 0, maxSteps: Int)(
      v: Vertex
  ): Iterable[(Vertex, Int)] =
    val (pos, dir, steps) = v
    val reachable = ArrayBuffer.empty[(Vertex, Int)]

    def addReachable(dir: Direction, newSteps: Int) =
      val (r, c) = dir + pos
      if r >= 0 && r < h && c >= 0 && c < w then
        reachable.addOne(((r, c), dir, newSteps), grid(r)(c))

    if steps < minSteps then
      addReachable(dir, steps + 1)
      return reachable

    if steps < maxSteps then addReachable(dir, steps + 1)
    addReachable(dir.cw, 1)
    addReachable(dir.ccw, 1)

    reachable
}

def dijkstra(
    s: Coord,
    t: Coord,
    adj: Vertex => IterableOnce[(Vertex, Int)]
): Int =
  val pq = PriorityQueue.empty[(Int, Vertex)](Ordering.by(-_._1))
  val dist = MutableMap.empty[Vertex, Int].withDefaultValue(Int.MaxValue)

  def add(v: Vertex, distance: Int) =
    if distance < dist(v) then
      dist.addOne(v, distance)
      pq.addOne((distance, v))

  add((s, (1, 0), 0), 0)
  add((s, (0, 1), 0), 0)

  while pq.nonEmpty do
    val (d, v) = pq.dequeue()
    if v._1 == t then return d
    else if d != dist(v) then ()
    else for (u, w) <- adj(v) do add(u, d + w)

  Int.MaxValue

extension (dir: Direction)
  def +(coord: Coord): Coord =
    (coord._1 + dir._1, coord._2 + dir._2)

  def cw = (dir._2, -dir._1)
  def ccw = (-dir._2, dir._1)

object Input {
  def read() =
    val grid = scala.io.Source.stdin.mkString.linesIterator
      .map(_.split("").map(_.toInt).toArray)
      .toArray
    Input(grid, grid(0).length, grid.length)
}

File: 2023/day17.py
from heapq import heappop as pop, heappush as push
from sys import stdin
from collections import defaultdict


grid = [[int(c) for c in line.strip()] for line in stdin]
w, h = len(grid[0]), len(grid)


def adj(v, part):
    (r, c), dir, steps = v
    reachable = []

    dcw, dccw = (dir[1], -dir[0]), (-dir[1], dir[0])
    left, right, straight = [(r + d[0], c + d[1]) for d in [dcw, dccw, dir]]

    if part == 2 and steps < 4:
        if straight[0] >= 0 and straight[0] < h and straight[1] >= 0 and straight[1] < w:
            reachable.append(((straight, dir, steps + 1),
                              grid[straight[0]][straight[1]]))
        return reachable

    max_steps = 3 if part == 1 else 10
    if steps < max_steps and straight[0] >= 0 and straight[0] < h and straight[1] >= 0 and straight[1] < w:
        reachable.append(((straight, dir, steps + 1),
                         grid[straight[0]][straight[1]]))
    if left[0] >= 0 and left[0] < h and left[1] >= 0 and left[1] < w:
        reachable.append(((left, dcw, 1), grid[left[0]][left[1]]))
    if right[0] >= 0 and right[0] < h and right[1] >= 0 and right[1] < w:
        reachable.append(((right, dccw, 1), grid[right[0]][right[1]]))

    return reachable


def dijkstra(S, T, part):
    INF = 10**18
    dist = defaultdict(lambda: INF)
    pq = []

    def add(i, dst):
        if dst < dist[i]:
            dist[i] = dst
            push(pq, (dst, i))
    add((S, (1, 0), 0), 0)
    add((S, (0, 1), 0), 0)

    while pq:
        D, i = pop(pq)
        if i[0] == T:
            return D
        if D != dist[i]:
            continue
        for j, w in adj(i, part=part):
            add(j, D + w)

    return dist[T]


if __name__ == '__main__':
    print("Part 1:", dijkstra((0, 0), (h-1, w-1), part=1))
    print("Part 1:", dijkstra((0, 0), (h-1, w-1), part=2))
